gadget.yaml: add system-recovery partition

mvo5 opened the pull request:
For UC20 we want to have a recovery partition. This adds to
the existing system-boot partition. The size is currently set
to 600Mb - this is a bit arbitrary and will allow us roughtly
2 recovery systems. We have the following sizes today:
 * core18:  54MB
 * snapd:   23MB
 * gadget:   1MB
 * kernel: 200MB
The recovery partition is setup after the system-boot
partition.

It also adds a initial grub.cfg which is very basic and
only supports a single recovery system at this point but
that recovery system can be selected via the grubenv:
* `snap_recovery_kernel`
* `snap_recovery_core`


Mostly a stub for now but it will allow us to start building a "pc" gadget in the "20/edge" channel.

mvo5 committed:
gadget.yaml: add system-recovery partition

For UC20 we want to have a recovery partition. This adds to
the existing system-boot partition. The size is currently set
to 600Mb - this is a bit arbitrary and will allow us roughtly
2 recovery systems. We have the following sizes today:
 * core18:  54MB
 * snapd:   23MB
 * gadget:   1MB
 * kernel: 200MB
The recovery partition is setup after the system-boot
partition.

It also adds a initial grub.cfg which is very basic and
only supports a single recovery system at this point but
that recovery system can be selected via the grubenv:
- `snap_recovery_kernel`
- `snap_recovery_core`
settings.

xnox commented on gadget.yaml:
```diff
@@ -26,3 +26,18 @@ volumes:
+      - name: Recovery
+        filesystem: vfat
+        # FIXME: what type to use?
+        type: EF,C12A7328-F81F-11D2-BA4B-00A0C93EC93B
```
Hmmm.... That is ESP type, and you appear to have ESP like content too. I don't believe it is allowed as per UEFI spec to have multiple ESP partitions.

Recovery must be booted from the system-boot partition.... Or you need chainload from there....

mvo5 replied:
Thanks, I was suspecting this, I will change things.

mvo5 replied:
I changed this to have a single ESP partition now - I think this makes it easier and more reliable. However in my reading I could not find anything about a limit of ESP partitions. I don't know if the boot order is defined if there are multiple ones so I think its wise to limit ourselfs to just one (especially since we don't need multiple ones).

mvo5 committed:
recovery.md: add recovery partition overview.

mvo5 committed:
improve recovery boot menu

pedronis commented on recovery-grub.cfg:
```diff
@@ -0,0 +1,28 @@
+
+for system in /system/*
+do
+  menuentry "Recovery $system" $system {
```
why the multiple recovery entries, wasn't the plan do the choosing from initramfs?

mvo5 replied:
Yes, I changed it to a single static entry.

pedronis commented on recovery.md:
```diff
@@ -0,0 +1,82 @@
+The revery partition contains the following file layout:
+
+/system/<name>/snaps/{base.snap,kernel.snap,other...}
+/system/<name>/assertions.txt
```
not a fan of the ".txt" extetnsion here, also unclear why we shouldn't support multiple streams

zyga replied:
I think this should be a assertions.d style directory.

pedronis commented on recovery.md:
```diff
@@ -0,0 +1,82 @@
+is big enough to contain a certain number of recovery "systems" that
+can then be used to recover or reinstall a broken system.
+
+A full "seed" to reinstall on a generic amd64 system is roughly 280MB
```
do we how big things like network manager etc are?  just 20mb for those seems tight?

mvo5 replied:
I increased this now.

mvo5 committed:
fix chainloading

zyga commented on gadget.yaml:
```diff
@@ -14,8 +14,23 @@ volumes:
+        # max volume label length: 11 chars
+        filesystem-label: sys-recover
+        # too big? too small?
+        size: 600M
```
There's a recommended size for windows systems that we should probably follow. I think it's around a GB now.

mvo5 replied:
Yeah, Samuele also commented on this. I increased the size for now.

mvo5 commented on grub.cfg-recovery:
```diff
@@ -0,0 +1,27 @@
+
+for system in /system/*
+do
+  menuentry "Recovery $system" $system {
```
Remove this and just do have "recover", "install" - each will boot into a "select" boot mode which allows to select what recovery/install system to boot from the initramfs. After the selection a grubenv setting is used to do a one-off boot into this selected recovery/install system.

mvo5 committed:
use static recovery grub entry

pedronis commented on recovery.md:
```diff
@@ -22,15 +22,18 @@ filesystem issues.
-us a static grub.cfg menu.
+us a static grub.cfg menu. The selection of the recovery system will
+happen at a later time from initramfs. To do this the recovery system
+will boot into a sepecial "select" mode and then the selection is set
```
special

pedronis commented on recovery.md:
```diff
@@ -44,7 +47,8 @@ account for now. This will change in a later revision of this doc.
 *** if so, chainboot into the system-boot partition
-*** if not, boot into selected bootmode, set snap_recovery_mode to "recovery" or "install"
+*** if not, boot into recovery bootmode, set snap_mode="recovery"
+**** later the initamfs will allow selecting different recovery systems
```
initramfs

pedronis commented on recovery.md:
```diff
@@ -22,15 +22,18 @@ filesystem issues.
-parition is FAT so we need to put the assertions in the "stream"
-format on disk. The assertions.txt must include exactly one
-model assertion.
+Where <name> is an encoded date/time like 20190521-1213. The parition
```
the partition

pedronis commented on recovery.md:
```diff
@@ -22,15 +22,18 @@ filesystem issues.
-model assertion.
+Where <name> is an encoded date/time like 20190521-1213. The parition
+is FAT so we need to put the assertions in the "stream" format on disk
+with a short name. The assertion files must include exactly one model
```
with short filenames. The assertion files all together must ...

pedronis reviewed (commented):
just some comments on recovery.md. maybe it needs  a spellcheck pass

mvo5 committed:
fix typos in recovery.md

mvo5 committed:
gadget.yaml: increase size of recovery to 1G

mvo5 committed:
gadget.yaml: increase size system-boot to 100Mb (for unpacked kernels)

mvo5 committed:
drop use of "regexp" module

mvo5 commented on recovery.md:
```diff
@@ -0,0 +1,82 @@
+
+The revery partition contains the following file layout:
+
+/system/<name>/snaps/{base.snap,kernel.snap,other...}
```
<name> should include the branch/model (directories) - we need to consider upper/lower. What to boot initially needs to be written by `snap prepare-image` in grubenv

pedronis replied:
I can double check what are the relevant regexps in snapd

pedronis replied:
* snap names should be like  `^[a-z0-9-]*[a-z][a-z0-9-]*$`  and at most 40 chars
* model names should be like `^[a-zA-Z0-9](?:-?[a-zA-Z0-9])*$` but we enforce lowercase on top of that, no size limits ATM afaict
* brand names  should be like `^(?:[a-z0-9A-Z]{32}|[-a-z0-9]{2,28})$`,  there sadly  uppercase can happen, same problem with snap-ids

mvo5 replied:
Uh, so in theory we might have "Brand" and "brand" and on vfat its the same - that is very unfortunate, I think that means we need to reconsider using branch/model in the directory.

pedronis replied:
yes, but the example you gave is misleading,  nice brand names are enforced lowercase (like "canonical"), it's the random ones (same as snap ids) that can have uppercase, we can't assume the brand will be nice, is things like CA5GLZgNQWPhspDQK63Er46Uxz2SO7ez that have uppercase in them, it's unlikely they will collide but not impossible

mvo5 committed:
add new "role: system-recovery"

mvo5 committed:
clarify how first-boot works

mvo5 committed:
update for latest changes

Update for latest changes in
  https://github.com/CanonicalLtd/ubuntu-image/pull/171
and
  https://github.com/snapcore/snapd/pull/6899

mvo5 committed:
Merge remote-tracking branch 'upstream/20' into 20

mvo5 commented:
This has not changed since a couple of days and we use it for our work to build images now. I would like to merge it into the *20* branch of the gadget and upload it to the "20/edge" channel for the "pc" snap in the store. But I need a :+1: review for this…

sil2100 reviewed (approved):
Ok, this looks generally good to me, but I don't know if the current recovery story changed from what I have heard in Capetown. I guess the sizes will be tweaked still, right?

cmatsuoka committed:
handle recover_reboot mode and add debug shell

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

mvo5 committed:
Merge pull request #1 from cmatsuoka/writable-ramdisk

handle recover_reboot mode and add debug shell

mvo5 committed:
snapcraft: move to base: core20

mvo5 committed:
grub.cfg-recovery: update for new /systems/snaps/$snap_name layout

In London we decided to use a new directory layout and put the
snaps into /snaps on the ubuntu-seed (recovery) partition. This
PR updates the grub config to use the right kernel.

vorlonofportland commented on Makefile:
```diff
@@ -69,4 +69,4 @@ all:
 install:
 	install -m 644 pc-boot.img pc-core.img shim.efi.signed grubx64.efi $(DESTDIR)/
-	install -m 644 grub.conf grub.cfg $(DESTDIR)/
+	install -m 644 grub.cfg-recovery grub.cfg-normal grub.conf $(DESTDIR)/
```
'grub.conf' is not a native name for the config file on grub's side, this name exists for compatibility with snapd's requirements.  If there are now two grub.cfg's, should the grub.conf symlink be dropped?

vorlonofportland commented on gadget.yaml:
```diff
@@ -14,15 +14,31 @@ volumes:
+        # UEFI will boot the ESP partition by default first
         type: EF,C12A7328-F81F-11D2-BA4B-00A0C93EC93B
+        # max volume label length: 11 chars
+        filesystem-label: sys-recover
```
Per London sprint, name should now be 'Ubuntu-seed' instead?

vorlonofportland commented on gadget.yaml:
```diff
@@ -14,15 +14,31 @@ volumes:
+      - name: System boot
+        type: 0B,EBD0A0A2-B9E5-4433-87C0-68B6B72699C7
         filesystem: vfat
         filesystem-label: system-boot
```
Per London: Ubuntu-boot

vorlonofportland commented on gadget.yaml:
```diff
@@ -14,15 +14,31 @@ volumes:
         filesystem: vfat
         filesystem-label: system-boot
-        size: 50M
+        size: 100M
```
I'm skeptical of this as a size for the partition given that we will need to store 2x vmlinuz (+ embedded initramfs) on this fs for the FDE case.

vorlonofportland commented on grub.cfg-normal:
```diff
@@ -37,6 +37,6 @@ set cmdline="root=LABEL=$label snap_core=$snap_core snap_kernel=$snap_kernel ro
     search --label $label --set=writable
     loopback loop ($writable)/system-data/var/lib/snapd/snaps/$snap_kernel
-    linux (loop)/kernel.img $cmdline
+    linux (loop)/kernel.img $cmdline systemd.debug-shell=1
```
This looks like an unrelated change that doesn't belong on the actual core20 branch.

vorlonofportland commented on grub.cfg-recovery:
```diff
@@ -0,0 +1,40 @@
+# FIXME: find a better way? this ensures that someone instructed snapd
+#        to go into "install" mode it actually does that
+if [ "$snap_mode" = "install" ]; then
+    set default=1
```
Discussed in London that these don't need to be menu entries at all: if the chooser interaction is expected to be entirely via the initramfs, this could execute the boot commands inline instead of driving a menu selection.

mvo5 committed:
many: use lodon sprint partiton names

During the London sprint we decided to use the following partition
names:
- ubuntu-seed (previously: sys-recover)
- ubuntu-boot (previously: system-boot)
- ubuntu-data (previously: writable)

This commit updates the code to use these names.

xnox commented:
There are two branches in the repo one is called `20` the other is called `uc20` can we just use one or the other? Or make server-side gitbranch reference link to make one point to the other? (not sure if github supports that).

vorlonofportland commented:
> There are two branches in the repo one is called `20` the other is called `uc20` can we just use one or the other? Or make server-side gitbranch reference link to make one point to the other? (not sure if github supports that).

The uc20 branch is new, and appears to be a copy of 20.  I guess this is based on the discussion about wanting consistent naming of uc20 branches across repos, and we should consider uc20 authoritative now?  (Thus this PR will eventually need to be retargeted)

vorlonofportland commented:
In fact, this PR appears to be superseded by https://github.com/snapcore/pc-amd64-gadget/pull/19.
